File: src/main/java/wav/demon/StatCraft/Commands/SC/SCEnderPearls.java
package wav.demon.StatCraft.Commands.SC;

import com.mysema.query.Tuple;
import com.mysema.query.sql.SQLQuery;
import org.bukkit.command.CommandSender;
import wav.demon.StatCraft.Commands.ResponseBuilder;
import wav.demon.StatCraft.Querydsl.QEnderPearls;
import wav.demon.StatCraft.Querydsl.QPlayers;
import wav.demon.StatCraft.StatCraft;
import wav.demon.StatCraft.Util;

import java.util.List;

public class SCEnderPearls extends SCTemplate {

    public SCEnderPearls(StatCraft plugin) {
        super(plugin);
        this.plugin.getBaseCommand().registerCommand("enderpearls", this);
    }

    @Override
    public boolean hasPermission(CommandSender sender, String[] args) {
        return sender.hasPermission("statcraft.user.enderpearls");
    }

    @Override
    public String playerStatResponse(String name) {
        try {
            int id = plugin.getDatabaseManager().getPlayerId(name);
            if (id < 0)
                throw new Exception();

            QEnderPearls e = QEnderPearls.enderPearls;
            SQLQuery query = plugin.getDatabaseManager().getNewQuery();

            Tuple tuple = query.from(e).where(e.id.eq(id)).uniqueResult(e.amount, e.distance, e.maxThrow);

            if (tuple == null)
                throw new Exception();

            Integer amount = tuple.get(e.amount);
            Integer distance = tuple.get(e.distance);
            Integer maxThrow = tuple.get(e.maxThrow);

            amount = amount == null ? 0 : amount;
            distance = distance == null ? 0 : distance;
            maxThrow = maxThrow == null ? 0 : maxThrow;

            return new ResponseBuilder(plugin)
                .setName(name)
                .setStatName("Ender Pearls Thrown")
                .addStat("Thrown", df.format(amount))
                .addStat("Distance", Util.distanceUnits(distance))
                .addStat("Farthest Throw", Util.distanceUnits(maxThrow))
                .toString();
        } catch (Exception ex) {
            return new ResponseBuilder(plugin)
                .setName(name)
                .setStatName("Ender Pearls Thrown")
                .addStat("Thrown", String.valueOf(0))
                .addStat("Distance", Util.distanceUnits(0))
                .addStat("Farthest Throw", Util.distanceUnits(0))
                .toString();
        }
    }

    @Override
    public String serverStatListResponse(int num) {
        QEnderPearls e = QEnderPearls.enderPearls;
        QPlayers p = QPlayers.players;
        SQLQuery query = plugin.getDatabaseManager().getNewQuery();

        List<Tuple> list = query
            .from(e)
            .leftJoin(p)
            .on(e.id.eq(p.id))
            .groupBy(p.name)
            .orderBy(e.amount.desc())
            .limit(num)
            .list(p.name, e.amount);

        return topListResponse("Ender Pearls Thrown", list);
    }
}

File: src/main/kotlin/com/demonwav/statcraft/commands/sc/SCEnderPearls.kt
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * https://www.demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.commands.sc

import com.demonwav.statcraft.StatCraft
import com.demonwav.statcraft.Util
import com.demonwav.statcraft.commands.ResponseBuilderKt
import com.demonwav.statcraft.commands.SecondaryArgument
import com.demonwav.statcraft.magic.ProjectilesCode
import com.demonwav.statcraft.querydsl.QPlayers
import com.demonwav.statcraft.querydsl.QProjectiles
import com.mysema.query.types.path.NumberPath
import org.bukkit.command.CommandSender
import java.sql.Connection
import java.util.LinkedList

class SCEnderPearls(plugin: StatCraft) : SCTemplate(plugin) {

    init {
        plugin.baseCommand.registerCommand("enderpearls", this)
    }

    override fun hasPermission(sender: CommandSender, args: Array<out String>?) = sender.hasPermission("statcraft.user.enderpearls")

    override fun playerStatResponse(name: String, args: List<String>, connection: Connection): String {
        try {
            val id = getId(name) ?: throw Exception()

            val p = QProjectiles.projectiles
            val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError

            val tuple = query.from(p).where(p.id.eq(id), p.type.eq(ProjectilesCode.ENDER_PEARL.code))
                .uniqueResult(p.amount.sum(), p.totalDistance.sum(), p.maxThrow.max()) ?: throw Exception()

            val amount = tuple.get(p.amount) ?: 0
            val distance = tuple.get(p.totalDistance) ?: 0
            val maxThrow = tuple.get(p.maxThrow) ?: 0

            return ResponseBuilderKt.build(plugin) {
                playerName { name }
                statName { "Ender Pearls Thrown" }
                stats["Total"] = df.format(amount)
                stats["Distance"] = Util.distanceUnits(distance)
                stats["Farthest Throw"] = Util.distanceUnits(maxThrow)
            }
        } catch (e: Exception) {
            return ResponseBuilderKt.build(plugin) {
                playerName { name }
                statName { "Ender Pearls Thrown" }
                stats["Thrown"] = 0.toString()
                stats["Distance"] = 0.toString()
                stats["Farthest Throw"] = 0.toString()
            }
        }
    }

    @SecondaryArgument("distance", "farthest")
    override fun serverStatListResponse(num: Long, args: List<String>, connection: Connection): String {
        var distance = false
        val p = QProjectiles.projectiles
        val pl = QPlayers.players

        val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError

        val path: NumberPath<Int>
        val title: String

        val arg = if (args.size > 0) {
            args[0]
        } else {
            ""
        }

        when (arg) {
            "distance" -> {
                path = p.totalDistance
                title = "Distance Traveled With Ender Pearls"
                distance = true
            }
            "farthest" -> {
                path = p.maxThrow
                title = "Farthest Ender Pearl Throw"
                distance = true
            }
            else -> {
                path = p.amount
                title = "Ender Pearls Thrown"
            }
        }

        val list = query
            .from(p)
            .leftJoin(pl)
            .on(p.id.eq(pl.id))
            .where(p.type.eq(ProjectilesCode.ENDER_PEARL.code))
            .groupBy(pl.name)
            .orderBy(path.desc())
            .limit(num)
            .list(pl.name, path)

        return if (distance) {
            topListDistanceResponse(title, list)
        } else {
            topListResponse(title, list)
        }
    }

    override fun onTabComplete(sender: CommandSender, args: Array<out String>): List<String> {
        if (args[args.size - 1].startsWith("-")) {
            var top = false
            for (s in args) {
                if (s.startsWith("-top")) {
                    top = true
                }
            }

            if (top) {
                val list = LinkedList<String>()
                list.add("-all")
                list.add("-distance")
                list.add("-farthest")

                return list.filter { it.startsWith(args[args.size - 1]) }
            } else {
                return super.onTabComplete(sender, args)
            }
        } else {
            return super.onTabComplete(sender, args)
        }
    }
}
